src/sources/alphavantage.rs

```diff
@@ -0,0 +1,152 @@
+//! Alphavantage Source Provider
+//! <https://www.alphavantage.co/>
+//!
+
+use super::{Pair, Price};
+use crate::error::Error;
+use bytes::buf::ext::BufExt;
+use hyper::{
+    client::{Client, HttpConnector},
+    header, Body, Request,
+};
+use hyper_rustls::HttpsConnector;
+use serde::{Deserialize, Serialize};
+
+/// Base URI for requests to the Coinone API
+pub const BASE_URI: &str = "https://www.alphavantage.co/";
+
+/// User-Agent to send in HTTP request
+pub const USER_AGENT: &str = "iqlusion delphi";
+
+/// Source provider for Alphavantage
+pub struct AlphavantageSource {
+    http_client: Client<HttpsConnector<HttpConnector>>,
+}
+
+///Parameters for queries
+pub struct AlphavantageParams {
+    function: String,
+    from_currency: String,
+    to_currency: String,
+    apikey: String,
+}
+
+impl AlphavantageParams {
+    ///Convert params into url query parameters
+    pub fn to_query_string(&self) -> String {
+        format!(
+            "function={}&from_currency={}&to_currency={}&apikey={}",
+            self.function, self.from_currency, self.to_currency, self.apikey
+        )
+    }
+}
+
+impl AlphavantageSource {
+    /// Create a new Alphavantage source provider
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            http_client: Client::builder().build(HttpsConnector::new()),
+        }
+    }
+
+    /// Get trading pairs
+    pub async fn trading_pairs(&self, pair: &Pair) -> Result<Response, Error> {
+        let params = AlphavantageParams {
+            function: "CURRENCY_EXCHANGE_RATE".to_owned(),
+            from_currency: pair.0.to_string(),
+            to_currency: pair.1.to_string(),
+            apikey: std::env::var("ALPHAVANTAGE_API").expect("Must set the API key"),
+        };
+
+        let uri = format!("{}query?{}", BASE_URI, params.to_query_string());
```

Comment:
If you commit the suggestion above, this becomes:

```suggestion
        let uri = params.to_request_uri());
```